ambuda/views/proofing/project.py

```diff
@@ -297,53 +398,207 @@ def replace(slug):
+                replaced_line = query_pattern.sub(replace, line)
                 matches.append(
                     {
-                        "query": escape(line).replace(
-                            query, Markup(f"<mark>{escape(query)}</mark>")
-                        ),
-                        "update": escape(line).replace(
-                            query, Markup(f"<mark>{escape(update)}</mark>")
-                        ),
+                        "query": line,
+                        "replace": replaced_line,
+                        "line_num": line_num,
                     }
                 )
-        if matches:
-            results.append(
-                {
-                    "slug": page_.slug,
-                    "matches": matches,
-                }
-            )
+
+        results.append({"page": page_, "matches": matches})
+        LOG.debug(f"{__name__}: Total matches appended: {len(matches)}")
+
+    selected_count = sum(value == "selected" for value in selected_keys.values())
+    LOG.debug(f"{__name__} > Number of selected changes = {selected_count}")
+
     return render_template(
-        "proofing/projects/replace.html",
+        "proofing/projects/confirm_changes.html",
         project=project_,
-        form=form,
+        form=ConfirmChangesForm(),
         query=query,
-        update=update,
+        replace=replace,
         results=results,
     )
 
 
+@bp.route("/<slug>/submit_changes", methods=["GET", "POST"])
+@login_required
+def submit_changes(slug):
+    """Submit selected changes across all of the project's pages.
+
+    This is useful to replace a string across the project in one shot.
+    """
+
+    project_ = q.project(slug)
+    if project_ is None:
+        abort(404)
+
+    LOG.debug(
+        f"{__name__}: SUBMIT_CHANGES --- {request.method} > {list(request.form.keys())}"
+    )
+
+    # FIXME: find a way to validate this form. Current `matches` are coming in the way of validators.
+    form = SubmitChangesForm(request.form)
+    # if not form.validate():
+    #     # elif request.form.get("form_submitted") is None:
+    #     invalid_keys = list(form.errors.keys())
+    #     LOG.debug(f'{__name__}: Invalid form values - {request.method}, invalid keys: {invalid_keys}')
+    #     return redirect(url_for("proofing.project.replace", slug=slug))
+
+    render = None
+    # search for "query" string and replace with "update" string
+    query = form.query.data
+    replace = form.replace.data
+
+    LOG.debug(
+        f"{__name__}: ({request.method})>  Got to submit method with {query}->{replace} "
+    )
+    LOG.debug(f"{__name__}: {request.method} > {list(request.form.keys())}")
+    selected_keys = {
+        key: value
+        for key, value in request.form.items()
+        if key.startswith("match") and not key.endswith("replace")
+    }
+    render = _select_changes(project_, selected_keys, query=query, replace=replace)
```

Comment:
There is more work to do. Ideally, pagination should be added. We need to catch runaway searches like "ॐ " or " the " so that memory consumption is bounded. I am riding on the fact that users are responsible.